packages/authentication/src/__tests__/fixtures/services/basic-auth-user-service.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/authentication
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {inject} from '@loopback/context';
+import {HttpErrors} from '@loopback/rest';
+import {UserService} from '../../../services/user.service';
+import {UserProfile} from '../../../types';
+import {USER_REPO} from '../keys';
+import {BasicAuthenticationStrategyCredentials} from '../strategies/basic-strategy';
+import {User} from '../users/user';
+import {UserRepository} from '../users/user.repository';
+
+export class BasicAuthenticationUserService
+  implements UserService<User, BasicAuthenticationStrategyCredentials> {
+  constructor(
+    @inject(USER_REPO)
+    private userRepository: UserRepository,
+  ) {}
+
+  async verifyCredentials(
+    credentials: BasicAuthenticationStrategyCredentials,
+  ): Promise<User> {
+    if (!credentials) {
+      throw new HttpErrors.Unauthorized(`'credentials' is null`);
+    }
+
+    if (!credentials.email) {
+      throw new HttpErrors.Unauthorized(`'credentials.email' is null`);
+    }
+
+    if (!credentials.password) {
+      throw new HttpErrors.Unauthorized(`'credentials.password' is null`);
+    }
+
+    const foundUser = this.userRepository.find(credentials.email);
+    if (!foundUser) {
+      throw new HttpErrors['Unauthorized'](
+        `User with email ${credentials.email} not found.`,
+      );
+    }
+
+    if (credentials.password !== foundUser.password) {
+      throw new HttpErrors.Unauthorized('The password is not correct.');
+    }
+
+    return foundUser;
+  }
+
+  convertToUserProfile(user: User): UserProfile {
+    if (!user) {
+      throw new HttpErrors.Unauthorized(`'user' is null`);
+    }
+
+    if (!user.id) {
+      throw new HttpErrors.Unauthorized(`'user.id' is null`);
+    }
+
+    if (!user.firstname) {
+      throw new HttpErrors.Unauthorized(`'user.firstname' is null`);
+    }
+
+    if (!user.surname) {
+      throw new HttpErrors.Unauthorized(`'user.surname' is null`);
+    }
+
+    if (!user.email) {
+      throw new HttpErrors.Unauthorized(`'user.email' is null`);
+    }
+
```

Comment:
Do we really want to make `firstname`, `surname`, `email` required? It is desired sometimes to hide fields such as `email` due to privacy.

I'm also not sure if these errors should be mapped to `HttpErrors.Unauthorized`.

Reply:
The token service and user service and strategies in src/__tests__/fixtures are mock implementations not production implementations in 'src'. Is it absolutely necessary for them to be perfect? The users must create their own production ready artifacts. (Just wondering)

Reply:
The code block above is called when the basic strategy is trying to verify that the user's credentials are good.

called from 
```ts
 async authenticate(request: Request): Promise<UserProfile | undefined> {
    const credentials: BasicAuthenticationStrategyCredentials = this.extractCredentals(
      request,
    );
    const user = await this.userService.verifyCredentials(credentials);
    const userProfile = this.userService.convertToUserProfile(user);

    return userProfile;
  }

```

Reply:
In this case, some of the user profile fields are not populated. It's not something wrong with the credentials. For basic auth, credentials are username/password.